client/main.rs

```diff
@@ -80,29 +100,64 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         params.throttle_parallel,
     );
 
-    let tasks: Vec<_> = params
-        .grpc_urls
-        .clone()
-        .iter_mut()
-        .map(|grpc_url| {
-            let grpc_parsed_url: Uri = grpc_url.parse().unwrap();
-
-            let tls_grpc_ca_cert = params.tls_grpc_ca_cert.clone();
-            let tls_grpc_client_key = params.tls_grpc_client_key.clone();
-            let tls_grpc_client_cert = params.tls_grpc_client_cert.clone();
-            let tx_transactions = tx_transactions.clone();
-
-            tokio::spawn(spawn_grpc_connection_with_retry(
-                grpc_parsed_url.clone(),
-                tls_grpc_ca_cert.clone(),
-                tls_grpc_client_key.clone(),
-                tls_grpc_client_cert.clone(),
-                tx_transactions.clone(),
-            ))
-        })
-        .collect();
-
-    futures::future::join_all(tasks).await;
+    let all_tasks: Arc<RwLock<HashMap<String, JoinHandle<()>>>> =
+        Arc::new(RwLock::new(HashMap::new()));
+
+    build_tasks(
+        grpc_urls_from_file.clone(),
+        &Params::from_args(),
+        tx_transactions.clone(),
+        Arc::clone(&all_tasks),
+    )
+    .await;
+
+    loop {
+        match grpc_urls_channel_receiver.recv().await {
+            Some(new_grpc_urls) => {
+                let urls_to_spawn: Vec<String> = new_grpc_urls
+                    .iter()
+                    .filter(|x| !grpc_urls_from_file.contains(x))
+                    .map(|x| x.to_string())
+                    .collect();
+
+                let urls_to_abort: Vec<String> = grpc_urls_from_file
+                    .iter()
+                    .filter(|x| !new_grpc_urls.contains(x))
+                    .map(|x| x.to_string())
+                    .collect();
+
+                for i in urls_to_abort {
+                    let mut all_tasks = all_tasks.write().await;
```

Comment:
it doesn't required to acquire write lock for every abort, better acquire once for all aborts